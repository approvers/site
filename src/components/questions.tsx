import { Container, Text, VStack } from "@chakra-ui/react";
import type { ReactNode } from "react";
import React from "react";

import { ExternalLink } from "./link";

const QA = ({ question, answer }: { question: ReactNode; answer: ReactNode }): JSX.Element => {
  return (
    <Container as="section">
      <Text>
        <Text as="span" mr={2} color="highlighted" fontSize="2xl">
          Q.
        </Text>
        {question}
      </Text>
      <Text>
        <Text as="span" mr={2} color="highlighted" fontSize="2xl">
          A.
        </Text>
        {answer}
      </Text>
    </Container>
  );
};

const questionAnswers = [
  {
    question: "限界開発鯖ってなに?",
    answer: "限界開発鯖は高専生を中心に限界になった人々が限界状態で開発を行う Discord 鯖です。",
  },
  {
    question: "限界開発鯖の雰囲気が知りたい！",
    answer: (
      <>
        こちらへどうぞ &rarr;{" "}
        <ExternalLink href="https://twitter.com/search?q=%23限界開発鯖&src=typed_query">
          #限界開発鯖
        </ExternalLink>
      </>
    ),
  },
  {
    question: "どうやって参加するの？",
    answer: (
      <>
        弊サーバーは、<Text as="b">完全紹介制</Text>です。
        知り合いのメンバーから招待リンクを受け取ってください。
      </>
    ),
  },
] as const;

export const Questions = (): JSX.Element => (
  <VStack pb={16}>
    {questionAnswers.map((props, i) => (
      <QA key={i} {...props} />
    ))}
  </VStack>
);
