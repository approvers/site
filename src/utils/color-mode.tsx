"use client";

import type { IconButtonProps, SpanProps } from "@chakra-ui/react";
import { ClientOnly, IconButton, Skeleton, Span } from "@chakra-ui/react";
import { ThemeProvider, useTheme } from "next-themes";
import type { ThemeProviderProps } from "next-themes";
import * as React from "react";
import { LuMoon, LuSun } from "react-icons/lu";

export interface ColorModeProviderProps extends ThemeProviderProps {}

export function ColorModeProvider(props: ColorModeProviderProps) {
  return <ThemeProvider attribute="class" disableTransitionOnChange {...props} />;
}

export type ColorMode = "light" | "dark";

export interface UseColorModeReturn {
  colorMode: ColorMode;
  setColorMode: (colorMode: ColorMode) => void;
  toggleColorMode: () => void;
}

export function useColorMode(): UseColorModeReturn {
  const { resolvedTheme, setTheme, forcedTheme } = useTheme();
  const colorMode = forcedTheme || resolvedTheme;
  const toggleColorMode = () => {
    setTheme(resolvedTheme === "dark" ? "light" : "dark");
  };
  return {
    colorMode: colorMode as ColorMode,
    setColorMode: setTheme,
    toggleColorMode,
  };
}

export function useColorModeValue<T>(light: T, dark: T) {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? dark : light;
}

export function ColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuMoon /> : <LuSun />;
}

export function OppositeColorModeIcon() {
  const { colorMode } = useColorMode();
  return colorMode === "dark" ? <LuSun /> : <LuMoon />;
}

interface ColorModeButtonProps extends Omit<IconButtonProps, "aria-label"> {}

export const ColorModeButton = React.forwardRef<HTMLButtonElement, ColorModeButtonProps>(
  function ColorModeButton(props, ref) {
    const { toggleColorMode } = useColorMode();
    return (
      <ClientOnly fallback={<Skeleton boxSize="9" />}>
        <IconButton
          ref={ref}
          aria-label="Toggle color mode"
          onClick={toggleColorMode}
          size="sm"
          variant="ghost"
          {...props}
          css={{
            _icon: {
              width: "5",
              height: "5",
            },
          }}
        >
          <ColorModeIcon />
        </IconButton>
      </ClientOnly>
    );
  },
);

export const LightMode = React.forwardRef<HTMLSpanElement, SpanProps>(
  function LightMode(props, ref) {
    return (
      <Span
        className="chakra-theme light"
        ref={ref}
        display="contents"
        color="fg"
        colorPalette="gray"
        colorScheme="light"
        {...props}
      />
    );
  },
);

export const DarkMode = React.forwardRef<HTMLSpanElement, SpanProps>(function DarkMode(props, ref) {
  return (
    <Span
      className="chakra-theme dark"
      ref={ref}
      display="contents"
      color="fg"
      colorPalette="gray"
      colorScheme="dark"
      {...props}
    />
  );
});
