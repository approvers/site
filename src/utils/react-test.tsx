import reactTestRenderer, { ReactTestRenderer, TestRendererOptions } from "react-test-renderer";
import { ReactElement } from "react";

export const render = (
  ui: ReactElement,
  options?: TestRendererOptions | undefined,
): ReactTestRenderer => reactTestRenderer.create(ui, options);
