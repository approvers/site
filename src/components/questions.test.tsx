import { expect, it } from "vitest";

import { render } from "../utils/react-test";
import { Questions } from "./questions";

it("renders correctly", () => {
  const tree = render(<Questions />);
  expect(tree.toJSON()).toMatchSnapshot();
});
