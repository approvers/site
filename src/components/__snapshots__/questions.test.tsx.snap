// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="chakra-stack css-t7gyxs"
>
  <div
    className="chakra-container css-0"
  >
    <p
      className="chakra-text css-0"
    >
      <span
        className="chakra-text css-1p0mlmm"
      >
        Q.
      </span>
      限界開発鯖ってなに?
    </p>
    <p
      className="chakra-text css-0"
    >
      <span
        className="chakra-text css-1p0mlmm"
      >
        A.
      </span>
      限界開発鯖は高専生を中心に限界になった人々が限界状態で開発を行う Discord 鯖です。
    </p>
  </div>
  <div
    className="chakra-container css-0"
  >
    <p
      className="chakra-text css-0"
    >
      <span
        className="chakra-text css-1p0mlmm"
      >
        Q.
      </span>
      限界開発鯖の雰囲気が知りたい！
    </p>
    <p
      className="chakra-text css-0"
    >
      <span
        className="chakra-text css-1p0mlmm"
      >
        A.
      </span>
      こちらへどうぞ →
       
      <a
        className="chakra-link css-sjclu7"
        href="https://twitter.com/search?q=%23限界開発鯖&src=typed_query"
        rel="noopener noreferrer"
        target="_blank"
      >
        #限界開発鯖
      </a>
    </p>
  </div>
  <div
    className="chakra-container css-0"
  >
    <p
      className="chakra-text css-0"
    >
      <span
        className="chakra-text css-1p0mlmm"
      >
        Q.
      </span>
      どうやって参加するの？
    </p>
    <p
      className="chakra-text css-0"
    >
      <span
        className="chakra-text css-1p0mlmm"
      >
        A.
      </span>
      弊サーバーは、
      <b
        className="chakra-text css-0"
      >
        完全紹介制
      </b>
      です。 知り合いのメンバーから招待リンクを受け取ってください。
    </p>
  </div>
</div>
`;
